Fixed booking slots (#873)
sharetribe/flex-docs: src/docs/ftw-availability/availability-management/index.md

---
title: Managing listing availability
slug: availability-management
updated: 2025-03-24
category: template-availability
ingress:
  Learn about timezones and availability management in the Template.
published: true
---

This article explains how listing availability is handled in the
Sharetribe Web Template, with a focus on technical details and
implementation. For a general overview of availability management, see
our
[Help Center](https://www.sharetribe.com/help/en/articles/8413212-how-availability-management-works).
See also the
[reference documentation for availability management](<(/references/availability)>).

## Availability plan types and timezone support

> The API supports creating availability plans with either type

Reviewer: ```suggestion
The Sharetribe APIs support creating availability plans with either type
```

`availabilityPlan/day` with day-level availability or type
`availabilityPlan/time` with time range level availability. When a user
creates a listing with the Sharetribe Web Template, the listing is
always created with type `availabilityPlan/time`, whether the calendar
booking type is daily, nightly, hourly or fixed.

Using the `availabilityPlan/time` type means that the template can
specify a certain time zone for the listing's availability plan, which
is asked when creating a bookable listing. The timezone is associated
with the availability plan and ensures that all booking times are
interpreted according to the local time of the listing, rather than the

> time zone of the customer making the booking.

Reviewer 2: ```suggestion
time zone of the customer making the booking - or UTC time zone, which `availabilityPlan/day` uses internally.
```

Author: @Gnito When is UTC used instead of customer time zone?

Reviewer 2: @shareoc to illustrate availabilityPlan/day a bit:

If someone tried to book a listing on New Zealand (2.4.2025 - 4.4.2025) on top of FTW-daily,
what was actually booked was 00:00 moments on UTC, which translates as
2.4.2025 klo 13.00.00 - 4.4.2025 klo 13.00.00 (local time is 13 hours ahead of UTC)

However, there is a tradeoff. In order to
[query listings](https://www.sharetribe.com/api-reference/marketplace.html#query-listings)
that use `availabilityPlan/time`, the `availability` parameter needs to
be set to `time-full` or `time-partial`. Queries using this
`availability` parameter do not support pagination.

By default, the template will display the first 24 results when using
the date filter (and as pagination is not supported, the results will be
limited to 24). This value can be modified in
[SearchPage.duck.js file](https://github.com/sharetribe/web-template/blob/main/src/containers/SearchPage/SearchPage.duck.js#L21),

> up to a 100.

Reviewer: ```suggestion
so you can allow displaying up to a 100 results.
```

![Marketplace payment flow](dates-query.png)

## Availability management

### Availability plan

For daily and nightly listings, the template simplifies availability
management by allowing the provider to select their daily availability.

> Under the hood, availability is automatically set from `00:00` to

Reviewer: Would it make sense to expand the "under the hood", by e.g. linking to the function(s) that do(es) this in the template, or then adding a few sentences on how this happens?

`00:00` (i.e., midnight to midnight), enabling bookings to span for
consecutive days.

For hourly and fixed booking length listings, the template exposes full
control over start and end times when adjusting availability entries.
These entries can be consecutive or non-consecutive and may have varying
seat counts. While it is technically possible to create a continuous
availability window from `00:00` to `00:00`, enabling bookings that span
entire days or overnight, the template does not have built-in support
for creating hourly bookings that span multiple calendar days.

### Availability exceptions

Providers can set availability exceptions on listings to override the
default plan for individual time ranges. Availability exceptions are
created based on the unit type:

- For daily bookings, exceptions are created per day, and the start and
  end dates can be the same date
- For nightly bookings, exceptions are created per night, and the end
  date needs to be at least one day later than the start date
- For hourly bookings and fixed-length bookings, exceptions are created
  with a start time and an end time, and the end time needs to be at

>   least 5 minutes after the start time (both start and end time minutes

Reviewer 2: To me, this feels that this tries to communicate 2 different things:
- Template: for hourly bookings and fixed-length bookings, start and end times need to match with sharp hours.
- API: in practice, the API endpoint supports shorter exceptions.
  - The smallest exception range is 5 minutes, but the general rule is that start and end times need to map with 5-minute intervals: 00, 05, 10, 15, 20, 25, 30, 35, 40, 45, 50, or 55.

>   least 5 minutes after the start time (both start and end time minutes

Author: @Gnito Double-checking, sharp hours means that value is a factor of an hour, e.g. 15, 20 or 30 minutes?

visible to the provider of the listing. When other users view the
listing’s availability on the listing page, they only see whether the
listing is available, but they cannot see whether the availability
results from the availability plan, exceptions, or existing bookings.

## Listing search and timezone coverage

The operator can allow users to search for listings based on their
availability on certain dates. The date search cannot be restricted to
specific time zones. This means that to cover listings from all possible
time zones, the template prolongs the start and end times before sending
them to the Sharetribe API. See the relevant code
[in SearchPage.duck.js](https://github.com/sharetribe/web-template/blob/main/src/containers/SearchPage/SearchPage.duck.js#L185-L186).

To be exact, the start time is moved 14 hours earlier, and the end time
is moved 12 hours later. This means that the availability search results
may include listings that are available before or after the time frame
that the user enters.

![Time range](time-range.png)

## Filtering timeslot with intervals

When querying timeslots for a listing using the booking unit "fixed" or
"hourly", the template uses

> [interval-based filtering](/references/availability). This approach

Reviewer: Should this point to `/references/availability/#interval-based-filtering` directly, or does it point to the whole article intentionally?

allows retrieving significantly less data from the API than using

> regular timeslot queries (the API is restricted to 500 timeslots per
> page). The query used in the template matches the first available

Reviewer 2: Should we give the context: pagination?

```suggestion
regular timeslot queries (the API is restricted to 500 timeslots per paginated
page). The query used in the template matches the first available
```

Reviewer: I would even expand the relevance of the pagination here:
- ...this approach allows retrieving significantly less data than regular time slot queries
- a regular timeslot query can fetch up to 500 time slots at once, and to fetch more you'd need to make multiple queries
- with short fixed time slots (e.g. 15 minutes), one day might have dozens of available time slots, so the first 500 time slots of a month might only cover a few weeks
- in the date picker, we're only interested in knowing whether there is availability on a certain day, so we make a query that matches the first available timeslot within the defined interval...

At least for me, it's not intuitive that 500 time slots is not enough for a monthly view of shorter time slots, so an explanation might make sense here

timeslot within the defined interval, and returns that. See the full
query in
[ListingPage.duck.js](https://github.com/sharetribe/web-template/blob/main/src/containers/ListingPage/ListingPage.duck.js#L495)
This information is used to display the availability of a listing in the
monthly calendar view when booking:

> ![Time range](april-availability.png)

Reviewer 2: I think the image should contain also dates that are not available!

For listings using the fixed booking slot unit type, the template
extends the end time of the timeslot query. If the duration of the fixed
booking slot is e.g. 2 hours, the end time of the query is extended by 2
hours. This allows for the timeslot query to return availability
correctly for the last moments of the month. For example, to display
availability on the last day of the month at 23:55 for a 2 hour fixed
booking slot, the template needs to check if there is availability 2
hours onwards from 23:55, therefore extending the query into the next
month. See the

> [relevant code in ListingPage.duck.js](https://github.com/sharetribe/web-template/blob/main/src/containers/ListingPage/ListingPage.duck.js#L488-L493).

Reviewer 2: I think this misses 1 paragraph that states how the date picker works with 'hour' and 'fixed' units. Something like this:

_So, the date picker fetches a reduced set of time slots for the calendar months if the unit type is 'hour' or 'fixed'. This means that, when a user selects one of the days with some availability on the date picker, the app needs to make another query to retrieve all the time slots available on that day._
